src/storage/src/source/timestamper.rs

```diff
@@ -0,0 +1,325 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Timestamper using persistent collection
+use std::collections::{HashMap, VecDeque};
+use std::task::{Context, Waker};
+
+use anyhow::Context as _;
+use futures::{FutureExt, StreamExt};
+use mz_dataflow_types::client::controller::storage::CollectionMetadata;
+use timely::progress::{Antichain, Timestamp as _};
+use timely::PartialOrder;
+
+use mz_dataflow_types::sources::MzOffset;
+use mz_expr::PartitionId;
+use mz_ore::now::NowFn;
+use mz_persist_client::read::{ListenEvent, ListenStream, ReadHandle};
+use mz_persist_client::write::WriteHandle;
+use mz_persist_client::Upper;
+use mz_repr::{Diff, Timestamp};
+
+pub struct CreateSourceTimestamper {
+    name: String,
+    read_progress: Antichain<Timestamp>,
+    write_upper: Antichain<Timestamp>,
+    persisted_timestamp_bindings: HashMap<PartitionId, VecDeque<(Timestamp, MzOffset)>>,
+    write_handle: WriteHandle<(), PartitionId, Timestamp, Diff>,
+    read_handle: ReadHandle<(), PartitionId, Timestamp, Diff>,
+    timestamp_bindings_listener: ListenStream<(), PartitionId, Timestamp, Diff>,
+    now: NowFn,
+    waker: Waker,
+}
+
+impl CreateSourceTimestamper {
+    pub async fn initialize(
+        name: String,
+        CollectionMetadata {
+            persist_location,
+            timestamp_shard_id,
+            tx_timestamp_shard_id: _,
+        }: CollectionMetadata,
+        now: NowFn,
+        waker: Waker,
+    ) -> anyhow::Result<Option<Self>> {
+        let persist_client = persist_location
+            .open()
+            .await
+            .with_context(|| "error creating persist client")?;
+
+        let mut persisted_timestamp_bindings = HashMap::new();
+
+        let (mut write_handle, read_handle) = persist_client
+            .open(timestamp_shard_id)
+            .await
+            .expect("persist handles open err");
+
+        // Collection is closed.  No source to run
+        if read_handle.since().is_empty() || write_handle.upper().is_empty() {
+            return Ok(None);
+        }
+
+        let read_progress = read_handle.since().clone();
+
+        // If this is the initialization of the persist collection, allow a read at Timestamp::minimum().  Else fetch
+        // the current upper.
+        let min_plus_one = Antichain::from_elem(Timestamp::minimum() + 1);
```

Comment:
Unconditionally advancing the timestamp is a bit strange, can we instead only take a snapshot when the upper is not zero? I can imagine using this reclock operator for reclocking data into other time domains, with a `NowFn` that could be an auto-increment value for example

Reply:
hmm sure -- I think we should just be able to initialize this with `write_handle.upper()`.  I think skipping snapshotting alone should work -- I'll check and make sure all the listens are happy being initialized at `Timestamp::minimum()`